### PR TITLE
Decrypt thread last reply on EventTile mount and update

### PR DESCRIPTION
**Pull request by gsouquet**
In conjuction with https://github.com/matrix-org/matrix-js-sdk/pull/2210, this fixes https://github.com/vector-im/element-web/issues/20709 & fixes https://github.com/vector-im/element-web/issues/21044

---
This PR currently has no changelog labels, so will not be included in changelogs.

Add one of: `T-Deprecation`, `T-Enhancement`, `T-Defect`, `T-Task` to indicate what type of change this is plus `X-Breaking-Change` if it's a breaking change.

Preview: https://pr7930--matrix-react-sdk.netlify.app
⚠️ Do you trust the author of this PR? Maybe this build will steal your keys or give you malware. Exercise caution. Use test accounts.

### CODE REVIEW

File: src/components/views/rooms/EventTile.tsx [hunk 1]
```diff
@@ -514,6 +514,13 @@ export default class EventTile extends React.Component<IProps, IState> {
 
         const room = MatrixClientPeg.get().getRoom(this.props.mxEvent.getRoomId());
         room?.on(ThreadEvent.New, this.onNewThread);
+
+        if (this.state.threadLastReply) {
+            if (this.state.threadLastReply.isEncrypted()) {
+                this.state.threadLastReply.once(MatrixEventEvent.Decrypted, this.onEventDecryption);
+                MatrixClientPeg.get().decryptEventIfNeeded(this.state.threadLastReply);
+            }
+        }
```
Comment (t3chguy) [inline src/components/views/rooms/EventTile.tsx:523]:
```suggestion
        if (this.state.threadLastReply?.isEncrypted()) {
            this.state.threadLastReply.once(MatrixEventEvent.Decrypted, this.onEventDecryption);
            MatrixClientPeg.get().decryptEventIfNeeded(this.state.threadLastReply);
        }
```

File: src/components/views/rooms/EventTile.tsx [hunk 2]
```diff
@@ -514,6 +514,13 @@ export default class EventTile extends React.Component<IProps, IState> {
 
         const room = MatrixClientPeg.get().getRoom(this.props.mxEvent.getRoomId());
         room?.on(ThreadEvent.New, this.onNewThread);
+
+        if (this.state.threadLastReply) {
+            if (this.state.threadLastReply.isEncrypted()) {
+                this.state.threadLastReply.once(MatrixEventEvent.Decrypted, this.onEventDecryption);
+                MatrixClientPeg.get().decryptEventIfNeeded(this.state.threadLastReply);
```
Comment (t3chguy) [inline src/components/views/rooms/EventTile.tsx:521]: This handler may never fire (and thus leak a bind reference) given that `decryptEventIfNeeded` does a bunch of short-circuits.

Can we use `decryptEventIfNeeded(...).then(this.onEventDecryption);`

File: src/components/views/rooms/EventTile.tsx [hunk 3]
```diff
@@ -596,6 +603,9 @@ export default class EventTile extends React.Component<IProps, IState> {
         if (this.threadState) {
             this.threadState.off(NotificationStateEvents.Update, this.onThreadStateUpdate);
         }
+        if (this.state.threadLastReply) {
+            this.state.threadLastReply.removeListener(MatrixEventEvent.Decrypted, this.onEventDecryption);
+        }
```
Comment (t3chguy) [inline src/components/views/rooms/EventTile.tsx:608]:
```suggestion
        this.state.threadLastReply?.removeListener(MatrixEventEvent.Decrypted, this.onEventDecryption);
```

File: src/components/views/rooms/EventTile.tsx [hunk 4]
```diff
@@ -604,8 +614,22 @@ export default class EventTile extends React.Component<IProps, IState> {
...
         }
+
+        if (this.state.threadLastReply !== prevState.threadLastReply) {
+            if (this.state.threadLastReply.isEncrypted()) {
+                this.state.threadLastReply.once(MatrixEventEvent.Decrypted, this.onEventDecryption);
+                MatrixClientPeg.get().decryptEventIfNeeded(this.state.threadLastReply);
+            }
+            if (prevState.threadLastReply) {
+                prevState.threadLastReply.removeListener(MatrixEventEvent.Decrypted, this.onEventDecryption);
+            }
```
Comment (t3chguy) [inline src/components/views/rooms/EventTile.tsx:625]:
```suggestion
            prevState.threadLastReply?.removeListener(MatrixEventEvent.Decrypted, this.onEventDecryption);
```

File: src/components/views/rooms/EventTile.tsx [hunk 5]
```diff
@@ -604,8 +614,22 @@ export default class EventTile extends React.Component<IProps, IState> {
             MatrixClientPeg.get().on(RoomEvent.Receipt, this.onRoomReceipt);
             this.isListeningForReceipts = true;
         }
+
+        if (this.state.threadLastReply !== prevState.threadLastReply) {
+            if (this.state.threadLastReply.isEncrypted()) {
+                this.state.threadLastReply.once(MatrixEventEvent.Decrypted, this.onEventDecryption);
+                MatrixClientPeg.get().decryptEventIfNeeded(this.state.threadLastReply);
```
Comment (t3chguy) [inline src/components/views/rooms/EventTile.tsx:621]: Same as above

**Review by t3chguy — APPROVED** [review]
Looks sane otherwise

File: src/components/views/rooms/EventTile.tsx (hunk 2 above)
Comment (gsouquet) [reply src/components/views/rooms/EventTile.tsx:521]: Hm but that event listener will be removed when the component unmount or when `threadLastReply` changes.
Happy to go with `.then()` but that will be impossible cancel the callback. If the component gets unmounted, and the event gets decrypted 1 second after we'll call `forceUpdate` on an unmounted component

**Review by t3chguy — APPROVED** [review]

**Pull request merged**